test/infrastructure/docker/controllers/dockermachine_controller.go

```diff
@@ -176,6 +176,19 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
 		role = constants.ControlPlaneNodeRoleValue
 	}
 
+	// Defining a cleanup func that will delete a machine when there are error during provisioning, so the operation
+	// can be re-tried from a clean state when the next reconcile happens (in 10 seconds)
+	defer func() {
+		if retErr != nil && !dockerMachine.Spec.Bootstrapped {
+			log.Info(fmt.Sprintf("%v, cleaning up so we can re-provision from a clean state", retErr))
+			if err := externalMachine.Delete(ctx); err != nil {
+				log.Info("Failed to cleanup machine")
+			}
+			res = ctrl.Result{RequeueAfter: 10 * time.Second}
+			retErr = nil
+		}
+	}()
```

Comment:
If I understand this correctly this would delete the underlying infrastructure regardless of the returned error if the actual docker machine was never bootstrapped. 

Should we try instead to add some retry logic in the container bootstrapping mechanism, or do we prefer to do it in the controller here?

Reply:
Looking at the error logs, when the container does not start properly we start getting weird errors like can't create pki folder, and those error does not go away after many retries :-(